Public API for animated icons, and package skeleton.

amirh opened the pull request:
This is just the skeleton for the animated icons package, a main design
goal is to allow us to keep the implementation in a private API and
still test it.

Tracking bug: https://github.com/flutter/flutter/issues/13345

amirh committed:
Public API for animated icons, and package skeleton.

This is just the skeleton for the animated icons package, a main design
goal is to allow us to keep the implementation in a private API and
still test it.

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons.dart:
```diff
@@ -0,0 +1,89 @@
+/// The available icons are specified in [AnimatedIcons].
+class AnimatedIcon extends StatelessWidget {
+
+  const AnimatedIcon({
```
run the analyzer with --dartdocs to make sure you catch all the places that need dartdocs, such as this constructor (which should, e.g., talk about how progress, color, and icon should not be null)

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons.dart:
```diff
@@ -0,0 +1,89 @@
+    this.semanticLabel,
+    this.textDirection,
+    // TODO(amirh): add a parameter for controlling scaling behavior.
+  }) : assert (progress != null),
```
nit: extraneous space after `assert`. (try running `dartfmt`? it might work well enough for this file)

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons.dart:
```diff
@@ -0,0 +1,89 @@
+       assert (icon != null);
+
+  /// The animation progress for the animated icon.
+  /// The value must be between 0 and 1.
```
You can't guarantee that the value will be between 0.0 and 1.0; if you don't support values outside that range, clamp them internally.

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons.dart:
```diff
@@ -0,0 +1,89 @@
+  @override
+  Widget build(BuildContext context) {
+
+    // TODO(amirh): implement semantics and text direction.
```
we should do that in this patch

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons_data.dart:
```diff
@@ -0,0 +1,21 @@
+part of material_animated_icons;
+
+class AnimatedIcons {
+  static const AnimatedIconData menu_arrow = const _MenuArrowIconData();
```
needs a private constructor like the Icons class

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons_private.dart:
```diff
@@ -0,0 +1,32 @@
+
+  final Animation<double> progress;
+  final Color color;
+  final AnimatedIconData icon;
```
since _AnimatedIcon is private, this could just refer to _AnimatedIconData, so it's clearer

Hixie commented on packages/flutter/lib/src/material_animated_icons/animated_icons_private.dart:
```diff
@@ -0,0 +1,32 @@
+
+part of material_animated_icons;
+
+class _AnimatedIcon extends StatelessWidget {
```
given the structure of this patch, i'd probably merge AnimatedIcon and _AnimatedIcon. The private class isn't giving you any value really.

amirh replied:
good point

Hixie commented on packages/flutter/lib/src/material_animated_icons/data/menu_arrow.g.dart:
```diff
@@ -0,0 +1,14 @@
+  const _MenuArrowIconData();
+
+  @override
+  String get placeHolderIconData => 'menu->arrow';
```
this should be a private member, if you want to prevent other classes from implementing it

Hixie commented:
This seems fine so far, but I think we should land it as a working unit (i.e. it should support animating at least one icon when it lands, with i18n and a11y etc). If you want to do review in increments, I recommend doing PRs against a branch on your clone, which can be reviewed incrementally.

amirh committed:
review comments

amirh commented:
Ok I'm going to be sending incremental PRs to amirh/flutter:animated-icons and once this is working merge back to flutter/flutter.

Addressed your comments and moved this PR to: https://github.com/amirh/flutter/pull/3/files
